src/ret_val.h

```diff
@@ -22,11 +22,13 @@ class ret_val
     public:
         ret_val() = delete;
 
-        static ret_val make_success( const std::string &msg, T val = default_success ) {
+        template<typename S, typename = typename std::enable_if< std::is_convertible<S, std::string>::value>::type>
+        static ret_val make_success( const S &msg, T val = default_success ) {
```

Comment:
You can still call this as `ret_val<bool>make_success( "a", "b")` - the `"a"` is convertible to a `std::string`, so it's enabled, but the `"b"` is also (implicitly!) converted to a `bool` for the second argument. You may even end up with the same problem in `ret_val<std::string>`.